### PR TITLE
ci: Add sync workflow

### PR DESCRIPTION
**Pull request by phlax**
Currently Envoy CI uses an SSH key with all powers to sync this repo by pushing to it

This is not optimally secure, and furthermore this pattern regularly causes Envoy CI to flake racing to checkout/push downstream repos.

I have created an app with just wf trigger permissions and added it to this repo

This PR adds a workflow that can be triggered by the app to sync Envoy by pulling from it

### CODE REVIEW

File: .github/workflows/envoy-sync.yaml
Comment (phlax) [inline .github/workflows/envoy-sync.yaml:22]: rm

**Comment by alecholmez** [comment]
@phlax let me know when this gets out of wip, thanks for this change this is great

**Comment by phlax** [comment]
>  let me know when this gets out of wip

will do, shouldnt take much, mostly i want to test it, but i think its also not quite right yet

**Comment by jpeach** [comment]
> I have created an app with just wf trigger permissions and added it to this repo
> 
> This PR adds a workflow that can be triggered by the app to sync Envoy by pulling from it

When this is ready, could you please also add an explanation (to the README maybe?) of where all the moving parts are, and how someone could update them? If the app has any source code, then it might be good for it to also live in the `envoyproxy` GitHub org.

File: README.md
Comment (phlax) [inline README.md:65]: my editor did a ws cleanup here

**Comment by phlax** [comment]
@alecholmez i have shifted the algo from the python script to here, i will update to use the envoy `api.go` target once it has been added (https://github.com/envoyproxy/envoy/pull/27675)

ive tested this pretty well now, so reasonably confident this should hit the ground working, but will follow up if there any issues

currently it takes ~3 mins, at the cost of some complexity it could be a lot faster with caching

> When this is ready, could you please also add an explanation (to the README maybe?)

note about sync added

> ... of where all the moving parts are, and how someone could update them?

they are here, pr to this repo

> If the app has any source code

it doesnt

> then it might be good for it to also live in the envoyproxy GitHub org.

it does

**Comment by phlax** [comment]
for ref, you can see a triggered wf run here https://github.com/phlax/go-control-plane/actions/runs/5111042721/jobs/9187570100

and a commit here https://github.com/phlax/go-control-plane/commit/ed3867062062a6a2b7f3f2ac0ab30480315b05c4

i have fixed the commit message

**Review by alecholmez — APPROVED** [review]
This looks great thanks for the addition

File: .github/workflows/envoy-sync.yaml
Comment (phlax) [reply .github/workflows/envoy-sync.yaml:15]: 20.04 is what is used/tested in envoy

Comment (alecholmez) [reply .github/workflows/envoy-sync.yaml:15]: Cool sounds good

**Comment by phlax** [comment]
can we land? i have the other repos that envoy syncs to updated, so we should be good to update envoy

File: ci/sync_envoy.sh
Comment (jpeach) [inline ci/sync_envoy.sh:31]: nit:
```
$ shellcheck ci/sync_envoy.sh

In ci/sync_envoy.sh line 31:
    read -a go_protos <<< "$(bazel query "kind('go_proto_library', ${GO_TARGETS[*]})" | tr '\n' ' ')"
    ^--^ SC2162 (info): read without -r will mangle backslashes.

For more information:
  https://www.shellcheck.net/wiki/SC2162 -- read without -r will mangle backs...
```

File: ci/sync_envoy.sh
Comment (jpeach) [inline ci/sync_envoy.sh:3]: nit: if you add `set -o errexit` here, it's the same as `bash -e`, except (IMHO) a bit clearer to the reader.

File: ci/sync_envoy.sh
Comment (jpeach) [inline ci/sync_envoy.sh:19]:
```suggestion
if [[ -z "$ENVOY_SRC_DIR" ]]; then
     echo "ENVOY_SRC_DIR not set, it should point to a cloned Envoy repo" >&2
     exit 1
 fi
 
 if [[ ! -d "$ENVOY_SRC_DIR" ]]; then
     echo "ENVOY_SRC_DIR ($ENVOY_SRC_DIR) not found, did you clone it?" >&2
     exit 1
 fi
```

File: ci/sync_envoy.sh
Comment (jpeach) [reply ci/sync_envoy.sh:3]: Also consider `set -o nounset`, which is also a useful check, though not always what you want.

File: README.md
Comment (jpeach) [inline README.md:13]: Nit:

```suggestion
The Go proto files are synced from the upstream Envoy repository (https://github.com/envoyproxy/envoy) on every upstream commit.
```

File: ci/sync_envoy.sh
Comment (jpeach) [inline ci/sync_envoy.sh:86]: Nit: is the `--global` flag needed?

**Review by jpeach — APPROVED** [review]
Nice! Added a few review comments, but just for tiny things. Not a big deal.

File: ci/sync_envoy.sh
Comment (phlax) [reply ci/sync_envoy.sh:3]: nouunset would not work here if it is checking for an unset var

**Comment by alecholmez** [comment]
@phlax do you have merge permissions? If not I can roll it in

File: ci/sync_envoy.sh
Comment (phlax) [reply ci/sync_envoy.sh:3]: > nit: if you add set -o errexit here, it's the same as bash -e, except (IMHO) a bit clearer to the reader.

thinking on this im not sure - my guess is that you if you surveyed devs more would recognize `-e` as its just far more common

re nounset - i guess it can probably work if i re/set it from the env

File: ci/sync_envoy.sh
Comment (phlax) [reply ci/sync_envoy.sh:19]: switched `-e` -> `-d`

File: ci/sync_envoy.sh
Comment (phlax) [reply ci/sync_envoy.sh:31]: would be good to add shellcheck to the ci here - i was running it locally before submitting but i think this slipped through

also, would be good to clear up the whitespace issues 

```console
$ pip install envoy.code.check
$ envoy.code.check . -c glint
```

will show you issues

File: ci/sync_envoy.sh
Comment (phlax) [reply ci/sync_envoy.sh:86]: no

**Comment by phlax** [comment]
> do you have merge permissions?

i do, but would rather let you guys do it

ive updated the pr with feedback

we can test it manually straight away when it lands, but i will also move to get the envoy side set up asap so we can test it properly - im happy to follow up if necessary

**Comment by alecholmez** [comment]
I'm good to merge this... @jpeach I'll defer to you clicking the green button since you had some comments

**Review by jpeach — APPROVED** [review]

**Comment by jpeach** [comment]
> I'm good to merge this... @jpeach I'll defer to you clicking the green button since you had some comments

Yeh, I just had a few nitpicks. This looks good overall 👍

**Comment by phlax** [comment]
after removing `nounset` (#706 ) manual run has worked (with no change)

```console
Syncing go protos ...
Copying ../envoy/build_go/envoy -> envoy
Copying ../envoy/build_go/contrib -> contrib
Committing changes ...
Nothing changed, not committing
```

https://github.com/envoyproxy/go-control-plane/actions/runs/5141777442/jobs/9254666886#step:5:160

i will land the update in Envoy shortly https://github.com/envoyproxy/envoy/pull/27294 and check from there

the real test will be when there is some change to commit, so will keep watching

**Comment by phlax** [comment]
looking good

image

**Pull request merged**
